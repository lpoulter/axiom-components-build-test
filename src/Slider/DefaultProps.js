export default {
  size: 'small',
  step: 1,
  valueFormatter: (n) => n,
  withTooltip: false,
};
