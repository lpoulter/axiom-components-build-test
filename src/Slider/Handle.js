import PropTypes from 'prop-types';
import React, { Component } from 'react';

import Position from '../Position/Position';
import PositionSource from '../Position/PositionSource';
import PositionTarget from '../Position/PositionTarget';
import TooltipContent from '../Tooltip/TooltipContent';
import TooltipContext from '../Tooltip/TooltipContext';

export default class Handle extends Component {
  static propTypes = {
    /** Disabled control of the Handle */
    disabled: PropTypes.bool,
    /** Visibility of the Handle */
    isVisible: PropTypes.bool,
    /** Callback for when the mouse is down **/
    onMouseDown: PropTypes.func,
    /** The width and height of the handle **/
    size: PropTypes.number,
    /** Value for the tooltip content */
    value: PropTypes.number,
    /** Value for the Handle position */
    valueAsPercentage: PropTypes.number,
    /** Value formatter for the tooltip */
    valueFormatter: PropTypes.func,
    /** Enables a value indicating tooltip */
    withTooltip: PropTypes.bool,
  };

  static defaultProps = {
    disabled: false,
    isVisible: true,
    onMouseDown: () => {},
    valueFormatter: (n) => n,
    withTooltip: false,
  };

  render() {

    const { disabled, onMouseDown, isVisible, value, valueAsPercentage, valueFormatter, withTooltip, size } = this.props;
    const setSize = size ? { width: `${size}px`, height:`${size}px` } : {};

    return (
      <Position
          enabled={ withTooltip }
          isVisible={ isVisible }
          showArrow>
        <PositionTarget>
          <div
              className="ax-slider__thumb"
              onMouseDown={ disabled ? null : onMouseDown }
              style={ { left: `${valueAsPercentage}%`, ...setSize } }/>
        </PositionTarget>

        <PositionSource>
          <TooltipContext theme="night" width="auto">
            <TooltipContent size="tiny" textStrong>
              { valueFormatter(value) }
            </TooltipContent>
          </TooltipContext>
        </PositionSource>
      </Position>
    );
  }
}
