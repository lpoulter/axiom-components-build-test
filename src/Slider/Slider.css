:root {
  --ax-slider-track-size--small: var(--spacing-grid-size--x1);
  --ax-slider-track-size--medium: var(--spacing-grid-size--x2);
  --ax-slider-thumb-size--small: var(--spacing-grid-size--x4);
  --ax-slider-thumb-size--medium: var(--spacing-grid-size--x6);
  --ax-slider-thumb-border-width: 0.0625rem;
}

.ax-slider {
  position: relative;
  outline: none;
  user-select: none;
}

.ax-slider__track {
  position: relative;
  top: 50%;
  right: 0;
  left: 0;
  transform: translateY(-50%);
  background-color: var(--color-theme-background--subtle);
  cursor: pointer;
}

.ax-slider__fill {
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  background-color: var(--color-ui-accent);
}

.ax-slider__marker {
  position: absolute;
  top: -50%;
  opacity: 0.75;
  background-color: var(--color-theme-background--overlay-heavy);
}

.ax-slider__thumb {
  position: absolute;
  top: 0;
  transform: translateX(-50%);
  border: var(--ax-slider-thumb-border-width) solid var(--color-theme-border);
  border-radius: 50%;
  box-shadow: var(--drop-shadow-theme-elevation--x1);
  background-color: var(--color-ui-white);
  cursor: pointer;
}

.ax-slider__range {
  & .ax-slider__thumb {
    transform: translateX(-50%) translateY(-30%);
  }
}

.ax-slider--small {
  height: var(--spacing-grid-size--x6);
  margin: 0 calc(var(--ax-slider-thumb-size--small) / 2);

  & .ax-slider__track,
  & .ax-slider__fill {
    height: var(--ax-slider-track-size--small);
    border-radius: var(--ax-slider-track-size--small);
  }

  & .ax-slider__marker {
    width: 0.125rem;
    height: 0.5rem;
  }

  & .ax-slider__thumb {
    width: var(--ax-slider-thumb-size--small);
    min-width: var(--spacing-grid-size--x3);
    height: var(--ax-slider-thumb-size--small);
    min-height: var(--spacing-grid-size--x3);
  }
}

.ax-slider--medium {
  height: var(--ax-slider-thumb-size--medium);
  margin: 0 calc(var(--ax-slider-thumb-size--medium) / 2);

  & .ax-slider__track,
  & .ax-slider__fill {
    height: var(--ax-slider-track-size--medium);
    border-radius: var(--ax-slider-track-size--medium);
  }

  & .ax-slider__marker {
    width: 0.25rem;
    height: 1rem;
  }

  & .ax-slider__thumb {
    width: var(--ax-slider-thumb-size--medium);
    min-width: var(--ax-slider-thumb-size--small);
    height: var(--ax-slider-thumb-size--medium);
    min-height: var(--ax-slider-thumb-size--small);
  }
}

.ax-slider--disabled {
  & .ax-slider__thumb {
    cursor: default;
  }

  & .ax-slider__fill {
    background-color: var(--color-theme-border);
  }
}

