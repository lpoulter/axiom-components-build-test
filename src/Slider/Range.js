import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import omit from 'lodash.omit';
import ResizeObserver from 'resize-observer-polyfill';
import Base from '../Base/Base';
import Grid from '../Grid/Grid';
import GridCell from '../Grid/GridCell';
import Handle from './Handle';
import Text from '../Typography/Text';
import sliderDefaultProps from './DefaultProps';
import './Slider.css';

const HIDE_LABELS_THRESHOLD = 150;
const SLIDER_HANDLE_SIZE = 0.01;

export default class Range extends Component {
  static propTypes = {
    /** Disabled control of the slider */
    disabled: PropTypes.bool,
    /** Value of the marker to indicate balanced position */
    markerValue: PropTypes.number,
    /** Maximum number that can be selected */
    max: PropTypes.number.isRequired,
    /** Maximum label */
    maxLabel: PropTypes.string,
    /** Minimum number that can be selected */
    min: PropTypes.number.isRequired,
    /** Minimum label */
    minLabel: PropTypes.string,
    /** Called when the sliders value changes */
    onChange: PropTypes.func.isRequired,
    /** Called when the slider is let go */
    onSlideEnd: PropTypes.func,
    /** Size of the slider */
    size: PropTypes.oneOf(['small', 'medium']),
    /** Configures the increments of the slide */
    step: PropTypes.number,
    /** Value formatter for the tooltip */
    valueFormatter: PropTypes.func,
    /** Values */
    values: PropTypes.arrayOf(PropTypes.number).isRequired,
    /** Enables a value indicating tooltip */
    withTooltip: PropTypes.bool,
  };

  static defaultProps = sliderDefaultProps;
  trackRef = React.createRef();
  rangeRef = React.createRef();

  constructor(props) {
    super(props);

    this.focusedHandleIndex = 0;

    this.state = {
      draggedHandleIndex: null,
      isMouseOver: false,
      showLabels: true,
      rangeWidth: '104',
    };

    this.ensureBoundsInRange();

    this.handleBlur = this.handleBlur.bind(this);
    this.handleFocus = this.handleFocus.bind(this);
    this.handleKeyDown = this.handleKeyDown.bind(this);
    this.handleMouseDown = this.handleMouseDown.bind(this);
    this.handleMouseMove = this.handleMouseMove.bind(this);
    this.handleMouseUp = this.handleMouseUp.bind(this);
    this.handleResize = this.handleResize.bind(this);

    this.observer = new ResizeObserver(([entry]) => this.handleResize(entry.contentRect.width));
  }


  ensureBoundsInRange() {
    const { values } = this.props;

    const valuesInRange = values.map(value => this.ensureValueInRange(value));

    if (values.some(value => !valuesInRange.includes(value))) {
      this.onChange(valuesInRange);
    }
  }

  ensureValueInRange(value) {
    const { min, max, step } = this.props;
    const valueInRange = Math.max(min, Math.min(value, max));
    return step * Math.round(valueInRange / step);
  }

  ensureValuesDontCrossover(values) {
    const orderedValues = [...values];

    if (orderedValues[0] > orderedValues[1]) {
      switch (this.focusedHandleIndex) {
      case 0:
        orderedValues[0] = orderedValues[1];
        break;
      case 1:
        orderedValues[1] = orderedValues[0];
        break;
      default:
        throw new Error(`Unknown handle ${this.focusedHandleIndex} was in focus: change is not applied`);
      }
    }
    return orderedValues;
  }

  getPercentageFromValue(value) {
    const { min, max } = this.props;
    const valueInRange = this.ensureValueInRange(value);
    return ((valueInRange - min) / (max - min)) * 100;
  }

  getValueFromPosition(clientX) {
    const { max, min } = this.props;
    const { left: posMin, right: posMax } = this.trackRef.current.getBoundingClientRect();

    const value = (clientX - posMin) * (max - min) / (posMax - posMin) + min;

    return this.ensureValueInRange(value);
  }

  getValues(clientX) {
    const { values } = this.props;

    return values.map((value, index) => {
      if (index === this.focusedHandleIndex) {
        return this.getValueFromPosition(clientX);
      }
      return value;
    });
  }

  getClosestHandleIndex(clientX) {
    const { values } = this.props;
    const xValue = this.getValueFromPosition(clientX);
    let closestDist;

    return values.reduce((closestIndex, currentValue, currentIndex) => {
      const currentDist = Math.abs(xValue - currentValue);
      if ( closestDist === undefined || currentDist < closestDist) {
        closestDist = currentDist;
        return currentIndex;
      }
      if (currentDist === closestDist && xValue > currentValue) {
        return currentIndex;
      }
      return closestIndex;
    }, 0);
  }

  componentDidMount(){
    this.rangeRef.current && this.observer.observe(this.rangeRef.current);
  }

  componentWillUnmount() {
    document.removeEventListener('mousemove', this.handleMouseMove);
    document.removeEventListener('mouseup', this.handleMouseUp);
    this.observer.unobserve(this.rangeRef.current);
  }

  handleMouseDown(event) {
    const { clientX } = event;
    this.focusedHandleIndex = this.getClosestHandleIndex(clientX);

    event.preventDefault();

    this.setState({ draggedHandleIndex: this.focusedHandleIndex });

    this.handleMouseMove(event);

    document.addEventListener('mousemove', this.handleMouseMove);
    document.addEventListener('mouseup', this.handleMouseUp);
  }

  handleMouseUp(event) {
    const { onSlideEnd } = this.props;
    const { clientX } = event;
    const values = this.getValues(clientX);
    const orderedValues = this.ensureValuesDontCrossover(values);

    this.setState({ draggedHandleIndex: null });
    onSlideEnd && onSlideEnd(orderedValues);

    document.removeEventListener('mousemove', this.handleMouseMove);
    document.removeEventListener('mouseup', this.handleMouseUp);
  }

  handleBlur() {
    document.removeEventListener('keydown', this.handleKeyDown);
  }

  handleFocus() {
    document.addEventListener('keydown', this.handleKeyDown);
  }

  handleMouseMove(event) {
    const { clientX } = event;

    this.onChange(this.getValues(clientX));
  }

  handleKeyDown(event) {
    const { step } = this.props;
    const { values } = this.state;

    switch (event.key) {
    case 'ArrowLeft':
    case 'ArrowDown':
      values[this.focusedHandleIndex] = this.ensureValueInRange(values[this.focusedHandleIndex] - step);
      break;
    case 'ArrowRight':
    case 'ArrowUp':
      values[this.focusedHandleIndex] = this.ensureValueInRange(values[this.focusedHandleIndex] + step);
      break;
    }

    this.onChange(values);
  }

  handleResize(rangeWidth) {
    this.setState({ rangeWidth });
  }

  onChange(values) {
    this.props.onChange(this.ensureValuesDontCrossover(values));
  }


  render() {
    const { disabled, markerValue, maxLabel, minLabel, size, values, valueFormatter, withTooltip, ...rest } = this.props;
    const { draggedHandleIndex, isMouseOver, rangeWidth } = this.state;
    const showLabels = rangeWidth >= HIDE_LABELS_THRESHOLD;
    const valuesAsPercentage = values.map(value => this.getPercentageFromValue(value));
    const classes = classnames('ax-slider', `ax-slider--${size}`, 'ax-slider__range', {
      'ax-slider--disabled': disabled,
    });

    const sliderFillStyle = { left:`${valuesAsPercentage[0]}%`, width: `${valuesAsPercentage[1] - valuesAsPercentage[0]}%` };

    const markerValueAsPercentage = this.getPercentageFromValue(markerValue);
    const sliderMarkerStyle = { left:`${markerValueAsPercentage}%` };

    return (
      <Grid baseRef={ this.rangeRef } horizontalGutters="tiny" responsive={ false }
          verticalAlign="middle" verticalGutters={ false } wrap={ false }>
        { showLabels &&
        <GridCell none={ true }>
          <Text>{ minLabel }</Text>
        </GridCell>
        }
        <GridCell width={ showLabels ? 70 : 100 }>
          <Base { ...omit(rest, ['onSlideEnd', 'min', 'max', 'step']) }
              className={ classes }
              onBlur={ this.handleBlur }
              onFocus={ disabled ? null : this.handleFocus }
              onMouseLeave={ () => this.setState({ isMouseOver: false }) }
              onMouseOver={ () => this.setState({ isMouseOver: true }) }
              tabIndex="0">
            <div
                className="ax-slider__track"
                onMouseDown={ disabled ? null : this.handleMouseDown }
                ref={ this.trackRef }>
              <div
                  className="ax-slider__fill"
                  style={ sliderFillStyle } />
              { markerValue !== undefined && <div
                  className="ax-slider__marker"
                  style={ sliderMarkerStyle } /> }
              <Handle disabled={ disabled } isVisible={ draggedHandleIndex === 0 || (draggedHandleIndex === null && isMouseOver) }
                  onMouseDown={ this.handleMouseDown }
                  size={ rangeWidth * SLIDER_HANDLE_SIZE }
                  value={ values[0] }
                  valueAsPercentage={ valuesAsPercentage[0] }
                  valueFormatter={ valueFormatter }
                  withTooltip={ withTooltip }/>
              <Handle disabled={ disabled } isVisible={ draggedHandleIndex === 1 || (draggedHandleIndex === null && isMouseOver) }
                  onMouseDown={ this.handleMouseDown }
                  size={ rangeWidth * SLIDER_HANDLE_SIZE }
                  value={ values[1] }
                  valueAsPercentage={ valuesAsPercentage[1] }
                  valueFormatter={ valueFormatter }
                  withTooltip={ withTooltip }/>
            </div>
          </Base>
        </GridCell>
        { showLabels &&
        <GridCell none={ true }>
          <Text>{ maxLabel }</Text>
        </GridCell>
        }
      </Grid>
    );
  }
}
