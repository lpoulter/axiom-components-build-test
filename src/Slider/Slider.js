import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import omit from 'lodash.omit';
import Base from '../Base/Base';
import Handle from './Handle';
import sliderDefaultProps from './DefaultProps';
import './Slider.css';

export default class Slider extends Component {
  static propTypes = {
    /** Disabled control of the slider */
    disabled: PropTypes.bool,
    /** Maximum number that can be selected */
    max: PropTypes.number.isRequired,
    /** Minimum number that can be selected */
    min: PropTypes.number.isRequired,
    /** Called when the sliders value changes */
    onChange: PropTypes.func.isRequired,
    /** Called when the slider is let go */
    onSlideEnd: PropTypes.func,
    /** Size of the slider */
    size: PropTypes.oneOf(['small', 'medium']),
    /** Configures the increments of the slide */
    step: PropTypes.number,
    /** Value */
    value: PropTypes.number.isRequired,
    /** Value formatter for the tooltip */
    valueFormatter: PropTypes.func,
    /** Enables a value indicating tooltip */
    withTooltip: PropTypes.bool,
  };


  static defaultProps = sliderDefaultProps;

  constructor(props) {
    super(props);
    this.state = {
      isDragging: false,
      isMouseOver: false,
    };

    this.handleBlur = this.handleBlur.bind(this);
    this.handleFocus = this.handleFocus.bind(this);
    this.handleKeyDown = this.handleKeyDown.bind(this);
    this.handleMouseDown = this.handleMouseDown.bind(this);
    this.handleMouseMove = this.handleMouseMove.bind(this);
    this.handleMouseUp = this.handleMouseUp.bind(this);
    this.getValue = this.getValue.bind(this);
  }

  getValue(clientX) {
    const { max, min, step } = this.props;

    const value = (clientX - this.posMin) * (max - min) / (this.posMax - this.posMin) + min;
    const roundedValue = step * Math.round(value / step);

    return Math.max(min, Math.min(roundedValue, max));
  }

  componentWillUnmount() {
    document.removeEventListener('mousemove', this.handleMouseMove);
    document.removeEventListener('mouseup', this.handleMouseUp);
  }

  handleMouseDown(event) {
    const { left, width } = this.container.getBoundingClientRect();

    event.preventDefault();
    this.posMin = left;
    this.posMax = left + width;
    this.handleMouseMove(event);
    this.setState({ isDragging: true });
    document.addEventListener('mousemove', this.handleMouseMove);
    document.addEventListener('mouseup', this.handleMouseUp);
  }

  handleMouseUp(event) {

    const { onSlideEnd } = this.props;
    const { clientX } = event;

    this.setState({ isDragging: false });
    onSlideEnd && onSlideEnd(this.getValue(clientX));

    document.removeEventListener('mousemove', this.handleMouseMove);
    document.removeEventListener('mouseup', this.handleMouseUp);
  }

  handleBlur() {
    document.removeEventListener('keydown', this.handleKeyDown);
  }

  handleFocus() {
    document.addEventListener('keydown', this.handleKeyDown);
  }

  handleMouseMove(event) {
    const { onChange } = this.props;
    const { clientX } = event;

    event.preventDefault();
    onChange(this.getValue(clientX));
  }

  handleKeyDown(event) {
    const { max, min, onChange, step, value } = this.props;

    switch (event.key) {
    case 'ArrowLeft':
    case 'ArrowDown':
      onChange(Math.max(min, Math.min(value - step, max)));
      break;
    case 'ArrowRight':
    case 'ArrowUp':
      onChange(Math.max(min, Math.min(value + step, max)));
      break;
    }
  }

  render() {
    const { disabled, max, min, size, value, valueFormatter, withTooltip, ...rest } = this.props;
    const { isDragging, isMouseOver } = this.state;
    const valueInRange = Math.max(min, Math.min(value, max));
    const valueAsPercentage = ((valueInRange - min) / (max - min)) * 100;
    const classes = classnames('ax-slider', `ax-slider--${size}`, {
      'ax-slider--disabled': disabled,
    });

    return (
      <Base { ...omit(rest, ['onSlideEnd', 'min', 'max', 'step']) }
          className={ classes }
          onBlur={ this.handleBlur }
          onFocus={ disabled ? null : this.handleFocus }
          onMouseLeave={ () => this.setState({ isMouseOver: false }) }
          onMouseOver={ () => this.setState({ isMouseOver: true }) }
          tabIndex="0">
        <div
            className="ax-slider__track"
            onMouseDown={ disabled ? null : this.handleMouseDown }
            ref={ (el) => this.container = el }>
          <div
              className="ax-slider__fill"
              style={ { width: `${valueAsPercentage}%` } } />
        </div>
        <Handle disabled={ disabled } isVisible={ isDragging || isMouseOver }
            onMouseDown={ this.handleMouseDown }
            value={ value } valueAsPercentage={ valueAsPercentage }
            valueFormatter={ valueFormatter }
            withTooltip={ withTooltip }/>
      </Base>
    );
  }
}
