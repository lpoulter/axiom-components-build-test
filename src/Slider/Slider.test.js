import React from 'react';
import renderer from 'react-test-renderer';
import Slider from './Slider';

const getComponent = (props = {}) => {
  const newProps = { disabled: false, min: 0, max: 20, value: 6.5, onChange: jest.fn(), ...props };
  return renderer.create(
    <Slider { ...newProps }/>
  );
};

describe('Slider', () => {
  it('renders with defaultProps', () => {
    const component = getComponent();
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders with custom props', () => {
    const component = getComponent({ size: 'medium', step: 0.1, valueFormatter: x=>Math.floor(x), withTooltip: true });
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
