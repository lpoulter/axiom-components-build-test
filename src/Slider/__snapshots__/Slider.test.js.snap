// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider renders with custom props 1`] = `
<div
  className="ax-slider ax-slider--medium"
  onBlur={[Function]}
  onChange={[MockFunction]}
  onFocus={[Function]}
  onMouseLeave={[Function]}
  onMouseOver={[Function]}
  tabIndex="0"
>
  <div
    className="ax-slider__track"
    onMouseDown={[Function]}
  >
    <div
      className="ax-slider__fill"
      style={
        Object {
          "width": "32.5%",
        }
      }
    />
  </div>
  <div
    className="ax-slider__thumb"
    onMouseDown={[Function]}
    style={
      Object {
        "left": "32.5%",
      }
    }
  />
</div>
`;

exports[`Slider renders with defaultProps 1`] = `
<div
  className="ax-slider ax-slider--small"
  onBlur={[Function]}
  onChange={[MockFunction]}
  onFocus={[Function]}
  onMouseLeave={[Function]}
  onMouseOver={[Function]}
  tabIndex="0"
>
  <div
    className="ax-slider__track"
    onMouseDown={[Function]}
  >
    <div
      className="ax-slider__fill"
      style={
        Object {
          "width": "32.5%",
        }
      }
    />
  </div>
  <div
    className="ax-slider__thumb"
    onMouseDown={[Function]}
    style={
      Object {
        "left": "32.5%",
      }
    }
  />
</div>
`;
